File: 0329. 矩阵中的最长递增路径/动态规划.swift
// 320ms
class Solution {
    func longestIncreasingPath(_ matrix: [[Int]]) -> Int {
        let row = matrix.count, col = matrix.first?.count ?? 0
        var table = [[Int]](repeating: [Int](repeating: -1, count: col), count: row)

        func memory(_ x: Int, _ y: Int) -> Int {
            if table[x][y] == - 1 {
                table[x][y] = search(x, y)
            }
            return table[x][y]
        }

        func search(_ x: Int, _ y: Int) -> Int {
            var result = 0
            if x > 0 && matrix[x][y] < matrix[x - 1][y] {
                result = max(result, memory(x - 1, y))
            }
            if x < row - 1 && matrix[x][y] < matrix[x + 1][y] {
                result = max(result, memory(x + 1, y))
            }
            if y > 0 && matrix[x][y] < matrix[x][y - 1] {
                result = max(result, memory(x, y - 1))
            }
            if y < col - 1 && matrix[x][y] < matrix[x][y + 1] {
                result = max(result, memory(x, y + 1))
            }
            return result
        }

        var result = 0
        for x in 0 ..< row {
            for y in 0 ..< col {
                result = max(maxn, memory(x, y))
            }
        }

        return result
    }
}

File: 1035. 不相交的线/动态规划.js
// 156ms
var maxUncrossedLines = function(nums1, nums2) {
    let row = nums1.length, col = nums2.length
    var table = new Array(row).fill(0).map(_ => new Array(col).fill(-1))

    function memory(i, j) {
        if (i < 0 || j < 0) {
            return 0
        } else if (table[i][j] === -1) {
            table[i][j] = search(i, j)
        }
        return table[i][j]
    }

    function search(i, j) {
        let result = 0
        if (nums1[i] == nums2[j]) {
            result = Math.max(result, memory(i-1, j-1) + 1)
        }
        return Math.max(result, memory(i, j-1), memory(i-1, j))
    }

    return memory(row-1, col-1)
};